batcheval: plumb TargetBytes/NumBytes into {Reverse,}Scan

tbg opened the pull request:
This hooks these new fields up to their already present counterparts in
the MVCC layer. This means that the KV server will now handle these
correctly.

The missing piece is to teach DistSender to keep track of the total
NumBytes seen across all ranges queried for the current batch, and to
terminate scans that have finished. This (and testing it) is left to a
follow-up PR.

Release note: None

tbg committed:
roachpb: add TargetBytes to batch header

Also adds NumBytes to the response header.

Release note: None

tbg committed:
batcheval: make Scan and ReverseScan twins

The code is now identical (mod a `Reverse: true`). It's tempting to merge them into one method,
but I assume there is a reason they're separated.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44925)
<!-- Reviewable:end -->

nvb reviewed (approved):
> The missing piece is to teach DistSender to keep track of the total
> NumBytes seen across all ranges queried for the current batch, and to
> terminate scans that have finished. This (and testing it) is left to a
> follow-up PR.

The other missing piece seems to be handling multiple scans in the same batch and adjusting their target bytes based on what is left over after the previous one has completed, right? We do this in both `evaluateBatch` and `DistSender.divideAndSendBatchToRanges`.

> Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44925#-:-M-kEMtu5f4uBnvhr81k:b-geaol2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/batcheval/cmd_scan_test.go, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44925#-M-kFMzFAn8o485-dhOV:-M-kFMzFAn8o485-dhOW:b-d8ct0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cb3e089386207aad795ed03cb28aaf37d79bb/pkg/storage/batcheval/cmd_scan_test.go#L32)):*
> ```Go
> 	const (
> 		tbNone = 0      // no limit, i.e. should return all kv pairs
> 		tbOne  = 1      // one byte = return first key only
> ```

I'd add a fourth test that consumes more than one key before hitting its limit but does actually hit a limit. Maybe write a third key and check that two are returned in this case?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> The other missing piece seems to be handling multiple scans in the same batch and adjusting their target bytes based on what is left over after the previous one has completed, right? We do this in both evaluateBatch and DistSender.divideAndSendBatchToRanges.

Yep, thanks for pointing that out. Updated commit message.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44925#-:-M-nWDyf0Pq1puycSh4m:bmgl260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/batcheval/cmd_scan_test.go, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44925#-M-kFMzFAn8o485-dhOV:-M-nVmWN7ct9z47yOwDb:bmubmu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cb3e089386207aad795ed03cb28aaf37d79bb/pkg/storage/batcheval/cmd_scan_test.go#L32)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd add a fourth test that consumes more than one key before hitting its limit but does actually hit a limit. Maybe write a third key and check that two are returned in this case?
</blockquote></details>

There's extensive datadriven testing for how much is actually returned in `mvcc_histories/target_bytes`. Here I'm mostly convincing myself that the plumbing is happening.


<!-- Sent from Reviewable.io -->

tbg committed:
batcheval: plumb TargetBytes/NumBytes into {Reverse,}Scan

This hooks these new fields up to their already present counterparts in
the MVCC layer. This means that the KV server will now handle these
correctly.

The missing pieces are two places where we need to keep track of the
NumBytes seen while handling a batch, one in `(*Replica).evaluateBatch`
and one in `(*DistSender).divideAndSendBatchToRanges`.

Release note: None

tbg commented:
bors r=nvanbenschoten
TFTR!
